server: added support for systemd socket activation and readiness reporting

pvaret opened the pull request:
This PR adds a command line flag to llama-server, `--systemd`, that makes it:

- use the socket passed by systemd, if any, instead of opening a new one on the given host and port;
- report readiness to serve queries using the systemd notification API.

This allows running llama-server as a systemd service, with on-demand activation, and reliable readiness reporting.

This PR makes use of macros to ensure that the systemd-related code is only built if the build is configured with `-DLLAMA_SERVER_SYSTEMD=ON`.

Per CONTRIBUTING.md, the modified files were formatted with clang-format. The formatting was applied in a distinct commit for easier reviewing.

pvaret committed:
Added support for systemd socket activation.

xuanson2025 commented:
Please revert the formatting commit. Any formatting that affect the whole file should be done in a dedicated PR.

pvaret committed:
Don't shut down the socket if it was passed down to us by systemd.
systemd owns the lifecycle of the socket, not us.

pvaret commented:
Sure -- done! Should I re-submit the formatting commit as a separate PR, or is that just not considered very valuable?

pvaret committed:
When using systemd, load the model before accepting connections.

This avoids a race condition where the client whose connection triggered
the service's start would get errors 503 while the model loads.

pvaret commented:
While waiting for someone to get around to this review, I made the following improvements:
- When using systemd socket activation, close() the socket on termination but don't use shutdown(), since we're not the socket owner.
- When using systemd socket activation, load the model before starting to accept() on the socket, in order to avoid a race condition where the client would get errors 503 while the model is loading.

Abrosimov-a-a commented:
To solve this task, you can use the following alternative hack:

```ini
[Service]
ExecStartPost=/bin/sh -c 'until curl -sf http://${HOST}:${PORT}/health | jq -e ".status == \\"ok\\"" > /dev/null; do sleep 0.2; done'
TimeoutStartSec=60
```

Make sure that `curl` and `jq` is installed on the system.
